src/cryptonote_core/blockchain.cpp

```diff
@@ -4300,40 +4360,67 @@ bool Blockchain::handle_block_to_main_chain(const block& bl, const crypto::hash&
 
 // XXX old code adds miner tx here
 
-  size_t tx_index = 0;
   // Iterate over the block's transaction hashes, grabbing each
-  // from the tx_pool and validating them.  Each is then added
+  // from the tx_pool (or from extra_block_txs) and validating them.  Each is then added
   // to txs.  Keys spent in each are added to <keys> by the double spend check.
   txs.reserve(bl.tx_hashes.size());
+  txs_meta.reserve(bl.tx_hashes.size());
+  txpool_events.reserve(bl.tx_hashes.size());
   for (const crypto::hash& tx_id : bl.tx_hashes)
   {
-    transaction tx_tmp;
-    blobdata txblob;
-    size_t tx_weight = 0;
-    uint64_t fee = 0;
-    bool relayed = false, do_not_relay = false, double_spend_seen = false, pruned = false;
     TIME_MEASURE_START(aa);
 
 // XXX old code does not check whether tx exists
     if (m_db->tx_exists(tx_id))
     {
       MERROR("Block with id: " << id << " attempting to add transaction already in blockchain with id: " << tx_id);
       bvc.m_verifivation_failed = true;
-      return_tx_to_pool(txs);
-      goto leave;
+      return_txs_to_pool();
+      return false;
     }
 
     TIME_MEASURE_FINISH(aa);
     t_exists += aa;
     TIME_MEASURE_START(bb);
 
-    // get transaction with hash <tx_id> from tx_pool
-    if(!m_tx_pool.take_tx(tx_id, tx_tmp, txblob, tx_weight, fee, relayed, do_not_relay, double_spend_seen, pruned))
+    // get transaction with hash <tx_id> from tx_pool or extra_block_txs
+    // other tx info we want:
+    //   * weight
+    //   * fee
+    //   * is pruned?
+    txs.emplace_back();
+    transaction &tx = txs.back().first;
+    blobdata &txblob = txs.back().second;
+
+    size_t tx_weight{};
+    uint64_t fee{};
+    bool pruned{};
+    bool found_tx_in_pool = true;
+
+    bool _unused1, _unused2, _unused3;
+    if(!m_tx_pool.take_tx(tx_id, tx, txblob, tx_weight, fee, _unused1, _unused2, _unused3, pruned))
```

Comment:
Nit: there could be another reason that `take_tx` returns false here besides that the tx doesn't exist (some other unexpected exception in `take_tx` after the tx meta *is* found in the pool).

Main reason I bring this up is because I think it would be nice to keep `MERROR("Failed to find tx_meta in txpool");`

I suggest something like:

```cpp
    {
      m_tx_pool.lock();
      bool failed = false;
      if (m_tx_pool.have_tx(tx_id, relay_category::all))
      {
        bool _unused1, _unused2, _unused3;
        bool r = m_tx_pool.take_tx(tx_id, tx, txblob, tx_weight, fee, _unused1, _unused2, _unused3, pruned);
        m_tx_pool.unlock();
        if (!r)
        {
          MERROR_VER("Block with id: " << id  << " failed to take tx: " << tx_id);
          failed = true;
        }
      }
      else
      {
        m_tx_pool.unlock();
        found_tx_in_pool = false;
        const auto extra_tx_it = extra_block_txs.txs_by_txid.find(tx_id);
        if (extra_tx_it != extra_block_txs.txs_by_txid.end()) // if txid exists in provided extra block txs
        {
          tx = std::move(extra_tx_it->second.first);
          txblob = std::move(extra_tx_it->second.second);
          tx_weight = get_transaction_weight(tx, txblob.size());
          fee = get_tx_fee(tx);
          pruned = tx.pruned;
          extra_block_txs.txs_by_txid.erase(extra_tx_it);
          txpool_events.emplace_back(txpool_event{tx, tx_id, txblob.size(), tx_weight, true});
        }
        else // did not find txid in mempool or provided extra block txs
        {
          MERROR_VER("Block with id: " << id  << " has at least one unknown transaction with id: " << tx_id);
          bvc.m_missing_txs = true;
          failed = true;
        }
      }

      if (failed)
      {
        txs.pop_back(); // We push to the back preemptively. On fail, we need txs & txs_meta to match size
        bvc.m_verifivation_failed = true;
        return_txs_to_pool();
        return false;
      }
    }
```

Up to you. I don't have a strong opinion.